doc/neps/nep-0057-array-in-array-out.rst

```diff
@@ -0,0 +1,231 @@
+replace immutable scalars are no longer hashable. For instance, they cannot
+by used directly as keys of dictionaries, the argument of an ``lru_cache``
+-decorated function, etc. In all circumstances, tbe patch is simple: convert
+the zero-rank array to a scalar with empty-tuple indexing.
+
+Running the test suites of dependent libraries against a branch of NumPy that
+implements these changes has revealed a few other issues. <Library maintainers,
+please summaries these issues here.>
+
+Detailed description
+--------------------
+
+The new functionality will be used in much the same was as old functionality,
+except that sometimes authors will need to convert zero-rank arrays to scalars
+rather than converting scalars to zero-rank arrays. For instance, consider the
+following:
+
+.. code:: python
+
+   import numpy as np
+   rng = np.random.default_rng(85878653462722874072976519960992129768)
+   x = rng.standard_normal(size=10)
+   y = np.sum(x, axis=-1)
+   z = {y: 'a duck'}  # use scalar result as dictionary key
+   y = np.asarray(y)  # convert to array to allow mutation
+   y[y < 0] = np.nan
+
+The use of ``x`` as a dictionary key would need to become ``x[()]``, but ``z``
+no longer needs to be explicitly converted to an array.
+
+.. code:: python
+
+   z = {y[()]: 'a duck'}  # extract scalar to use as dictionary key
+   y[y < 0] = np.nan      # no conversion to array required
+
+Realistic examples can be found throughout the codebases of dependent
+libraries.
+
+Related work
+------------
+
+All known libraries that attempt to implement the array API standard
+(e.g. ``cupy``, ``torch``, ``jax.numpy``, ``dask.array``) return
+zero-rank arrays as specified by the standard.
+
+Implementation
+--------------
+
+To implement the NEP, the branch prepared by Sebastian needs to be merged,
+and dependent libraries will need to prepare releases that adapt to (and take
+advantage of) the changes. Branches of libraries including SciPy and Matplotlib
+have already been prepared without much difficulty. Initially, users will have
+the option of opting into this behavior using an environment variable, so these
+releases need to be compatible with both old and new behaviors. To make the new
+behavior the default and only behavior, NumPy will need to advise users of the
+pending change and give the appropriate notice. The initial draft of this
+document does not specify the appropriate timeline and procedures; this line
+will be updated to reflect the consensus of the maintainer team.
+
+Alternatives
+------------
+
+There are two main alternatives to this proposal.
+
+The alternative suggested by :ref:`NEP56` is to maintain the current behavior
+and make NumPy scalars more fully duck-type zero-rank arrays, such as adding
+missing behaviors.
+
+While this work would still be valuable, we propose the behavior change to more
+fully comply with the standard and to resolve the problems mentioned in the
```

Comment:
I think this somewhat leaves out a huge part of it which is the non-array API part?  The whole "just add methods" works very well for array-api numeric types, but badly for general scalars like object/string dtypes.

Reply:
> I think this somewhat leaves out a huge part of it which is the non-array API part

Could you be more specific about what the "non-array API part" is? From context, I think you are referring to an important argument in favor of the change, but I don't know what argument that is.

Reply:
Could you reply to this part @seberg?

Reply:
I think our discussion above covered this, but I'll try to see if there is more (it may make sense to look at NumPy bugs around this, but SciPy fixes seem actually more interesting for the NEP).
I.e. I mainly mean the fact that sometimes you have to sprinkle in `np.asarray()` because of the magic 0-D -> scalar conversion.